zerver/webhooks/facebook/view.py

```diff
@@ -0,0 +1,123 @@
+                                _page_id=payload['entry']['changes']['value']['page_id'],
+                                _thread_id=payload['entry']['changes']['value']['thread_id']
+                            )
+    elif field == 'website':
+        message = message + 'Changed **website**'
+    return message
+
+def permissions_object(payload: Dict[Text, Any]) -> Text:
+    field = payload['entry']['changes']['field']
+    message = '**{_field} permission** changed'.format(_field=field)
+    if field == 'ads_management':
+        message = message + '\n* {_verb}'.format(
+            _verb=payload['entry']['changes']['value']['verb']
+        )
+        for id in payload['entry']['changes']['value']['target_ids']:
+            message = message + '\n  * {_id}'.format(_id=id)
+    elif field == 'manage_pages':
+        message = message + '\n* {_verb}'.format(
+            _verb=payload['entry']['changes']['value']['verb']
+        )
+        for id in payload['entry']['changes']['value']['target_ids']:
+            message = message + '\n  * {_id}'.format(_id=id)
+    return message
+
+def application_object(payload: Dict[Text, Any]) -> Text:
+    field = payload['entry']['changes']['field']
+    message = '**{_field}** received'.format(_field=field)
+    if field == 'plugin_comment':
+        message = message + '\n**{_msg_user}:**\n```quote\n{_message}\n```'.format(
+            _msg_user=payload['entry']['changes']['value']['from']['name'],
+            _message=payload['entry']['changes']['value']['message']
+        )
+    if field == 'plugin_comment_reply':
+        message = message + '\n**{_prt_msg_user}:** (Parent)\n' \
+                '```quote\n{_prt_message}\n```'.format(
+                    _prt_msg_user=payload['entry']['changes']['value']['parent']['from']['name'],
+                    _prt_message=payload['entry']['changes']['value']['parent']['message']
+                )
+        message = message + '\n**{_cld_msg_user}:**\n```quote\n' \
+                '```quote\n{_cld_message}\n```\n```'.format(
+                    _cld_msg_user=payload['entry']['changes']['value']['from']['name'],
+                    _cld_message=payload['entry']['changes']['value']['message']
+                )
+    return message
+
+@api_key_only_webhook_view("Facebook")
+@has_request_variables
+def api_facebook_webhook(request: HttpRequest, user_profile: UserProfile,
+                        stream: Text=REQ(default='Facebook'), token: Text=REQ(),
+                        payload: Dict[Text, Any]=REQ(argument_type='body')) -> HttpResponse:
+    tmp_body = str(payload)
+    tmp_body = tmp_body.replace("\'", '\"')
+    tmp_body = tmp_body.replace('["', '!#!$!%!&!')
+    tmp_body = tmp_body.replace('"]', '!&!%!$!#!')
+    tmp_body = tmp_body.replace('[', '')
+    tmp_body = tmp_body.replace(']', '')
+    tmp_body = tmp_body.replace("!#!$!%!&!", '["')
+    tmp_body = tmp_body.replace("!&!%!$!#!", '"]')
+    _payload = json.loads(tmp_body)
```

Comment:
Again, I am not convinced that you need all of this, your payloads look fine to me and shouldn't need this code to sanitize the JSON. Could you explain further as to why you are doing this?

Reply:
Removing this method like [view.py on my Gist](https://gist.github.com/nyan-salmon/271c7a67ec606ec33798f1de8f5d7905) gives the following error.

Cannot purse as `payload['entry']['changes']['value']['message']` because it is inserted in an unnecessary array.

```
(zulip-py3-venv) ~/zulip2 (master)$ ./tools/test-backend zerver/webhooks/facebook
-- Running tests in serial mode.
Destroying test database for alias 'default'...
Using existing clone for alias 'default'...
Running zerver.webhooks.facebook.tests.FacebookTests.test_application_plugin_comment
2018-01-11 15:22:08.949 ERR  [] Traceback (most recent call last):
  File "/srv/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/zulip-py3-venv/lib/python3.5/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/lib/request.py", line 170, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/decorator.py", line 306, in _wrapped_func_arguments
    raise err
  File "/home/zulipdev/zulip2/zerver/decorator.py", line 280, in _wrapped_func_arguments
    return view_func(request, user_profile, *args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/lib/request.py", line 170, in _wrapped_view_func
    return view_func(request, *args, **kwargs)
  File "/home/zulipdev/zulip2/zerver/webhooks/facebook/view.py", line 104, in api_facebook_webhook
    body = get_body_based_on_object(object)(_payload)
  File "/home/zulipdev/zulip2/zerver/webhooks/facebook/view.py", line 58, in application_object
    field = payload['entry']['changes']['field']
TypeError: list indices must be integers or slices, not str


======================================================================
FAIL: test_application_plugin_comment (zerver.webhooks.facebook.tests.FacebookTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/zulipdev/zulip2/zerver/webhooks/facebook/tests.py", line 18, in test_application_plugin_comment
    expected_subject, expected_message)
  File "/home/zulipdev/zulip2/zerver/lib/test_classes.py", line 633, in send_and_test_stream_message
    self.STREAM_NAME, **kwargs)
  File "/home/zulipdev/zulip2/zerver/lib/test_classes.py", line 573, in send_json_payload
    self.assert_json_success(result)
  File "/home/zulipdev/zulip2/zerver/lib/test_classes.py", line 449, in assert_json_success
    self.assertEqual(result.status_code, 200, json['msg'])
AssertionError: 500 != 200 : Internal server error

----------------------------------------------------------------------
Ran 1 test in 0.253s

FAILED (failures=1)
Destroying test database for alias 'default'...
```

Reply:
I wrote that you had to delete unnecessary arrays earlier, however for example, the target_ids array included in `fixtures/permissions_ads_management.json` is necessary.

So, the code as shown below doesn't work parfectly.
```
tmp_body = str(payload)
tmp_body = tmp_body.replace('[', '')
tmp_body = tmp_body.replace(']', '')
```

To protect a necessary array, We need the code as I commited.

Reply:
Okay, there are a couple of points I'd like you to consider:

* Converting a payload dictionary to a `str` and then loading it again with `json.loads` is not right, so remove all of that code and the `tmp_body.replace` lines.
* Once you remove all of that code and run into that error, try to fix the error by not changing the `payload` but by figuring out how you can get to the data you want. So the error you showed me seems to arise when you try to give a key where it expects a list index, i.e., a number, so look at the payload and see how you can get to the value you want using an index without removing the "unnecessary array".

A good thing to keep in mind is that you should never modify a payload's content, you should just access it and use the data in whatever way you want. :)